serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -90,22 +92,60 @@ def set_size(self, request_ids: List[int], target_sizes: List[int]):
+                del self.kv_cache.slot_mappings[id]
+
+            elif id in self.kv_cache.block_tables:
+                # Decoding
 
-            elif (
-                id in self.kv_cache.block_tables
-                and len(self.kv_cache.block_tables[id]) < num_needed_block
-            ):
-                # Decoding, need to allocate a new block for this request
-                assert len(self.kv_cache.block_tables[id]) + 1 == num_needed_block
-                self.kv_cache.block_tables[id].append(self.free_blocks.pop())
+                if len(self.kv_cache.block_tables[id]) < num_needed_block:
+                    # Need to allocate a new block for this request
+                    assert len(self.kv_cache.block_tables[id]) + 1 == num_needed_block
+                    self.kv_cache.block_tables[id].append(self.free_blocks.pop())
+
+                pos = size - 1
+                block_number = self.kv_cache.block_tables[id][-1]
+
+                if self.block_sliding_window:
+                    block_number = self.kv_cache.block_tables[id][
+                        (pos // self.block_size) % self.block_sliding_window
+                    ]
+                else:
+                    block_number = self.kv_cache.block_tables[id][-1]
+
+                block_offset = pos % self.block_size
+                slot = block_number * self.block_size + block_offset
+                self.kv_cache.slot_mappings[id].append(slot)
 
             elif id not in self.kv_cache.block_tables:
                 assert (
                     len(self.free_blocks) >= num_needed_block
                 ), "Not enough free blocks."
 
                 for _ in range(num_needed_block):
-                    self.kv_cache.block_tables[id].append(self.free_blocks.pop())
+                    block_number = self.free_blocks.pop()
+                    self.kv_cache.block_tables[id].append(block_number)
+
+                for block_idx in range(math.floor(size / self.block_size)):
+                    if self.block_sliding_window:
+                        block_idx %= self.block_sliding_window
+
+                    block_number = self.kv_cache.block_tables[id][block_idx]
+                    slots = [
+                        block_number * self.block_size + block_offset
+                        for block_offset in range(self.block_size)
+                    ]
+                    self.kv_cache.slot_mappings[id] += slots
+
+                for i in range(len(self.kv_cache.slot_mappings[id]), size):
+                    block_idx = i // self.block_size
+
+                    if self.block_sliding_window:
+                        block_idx %= self.block_sliding_window
+
+                    block_number = self.kv_cache.block_tables[id][block_idx]
+                    block_offset = i % self.block_size
+                    slot = block_number * self.block_size + block_offset
+                    self.kv_cache.slot_mappings[id].append(slot)
```

Comment:
@yelite What do you think about this "optimization"? It micro-optimizes the previous sequential loop over the prompt tokens by calculating `slots` one block at a time. But I'm not sure if this is worth it at the cost of the code complexity. So if you prefer a simpler naive loop I'll revert to the original code.